### PR TITLE
Add formatting

### PR DESCRIPTION
**Pull request by alex-schutz**
Making a pull request so that it's easier to comment on the code

### CODE REVIEW

File: .gitignore
Comment (alex-schutz) [inline .gitignore:1]: I've added a gitignore so the build files don't accidentally get added

File: Unit_Tests/test_CTP_generator.py
Comment (alex-schutz) [inline Unit_Tests/test_CTP_generator.py:1]: I recommend using a testing paradigm like [pytest](https://docs.pytest.org/en/stable/) to write your unit tests. It makes testing small bits of code much easier and does nice printing of the output so you can tell where things have failed. It'll also get picked up by the test section in VS Code so you can easily run all the tests/subsets of tests. 

If you ask ChatGPT to refactor these tests into a pytest format I expect it'll do a reasonable job. It's generally fairly decent at writing (simple) unit tests, but definitely check them over!

File: Unit_Tests/test_CTP_generator.py [hunk 1]
```diff
@@ -1,41 +1,58 @@
...
+        help="Number of nodes in the graph",
+        required=False,
+        default=5,
+    )
+    parser.add_argument(
+        "--directory",
+        type=str,
+        help="Directory to save results",
+        required=False,
+        default="C:\\Users\\shala\\Documents\\Oxford Undergrad\\4th Year\\4YP\\Code\\MARL_CTP\\Logs",
```
Comment (alex-schutz) [inline Unit_Tests/test_CTP_generator.py:28]: Something like `os.getcwd()` with `os.path.join` would be more portable

File: Unit_Tests/test_CTP_generator.py [hunk 2]
```diff
@@ -1,41 +1,58 @@
 import jax
 import jax.numpy as jnp
 import sys
-sys.path.append('..')
+
+sys.path.append("..")
```
Comment (alex-schutz) [inline Unit_Tests/test_CTP_generator.py:5]: Ideally you can avoid doing this by making the project into a package. Might require a small install script (but you should add one anyway to handle the `requirements.txt`)

**Review by alex-schutz — COMMENTED** [review]
I'm adding some comments on general python project management for now, and I'll get into the technical details later.

I've applied Black formatting to the python files, if you set your editor to do formatting on save it will mean you don't have to think about it. Autoformatting is really helpful when using git to do pull requests as it makes it much easier to compare what gets changed.

I would recommend adding a `requirements.txt` file which contains things like
```
absl-py==2.1.0
chex==0.1.87
contourpy==1.3.0
cycler==0.12.1
etils==1.10.0
flax==0.10.0
fonttools==4.54.1
fsspec==2024.10.0
humanize==4.11.0
importlib_resources==6.4.5
jax==0.4.35
jaxlib==0.4.35
jraph==0.0.6.dev0
kiwisolver==1.4.7
markdown-it-py==3.0.0
matplotlib==3.9.2
mdurl==0.1.2
ml_dtypes==0.5.0
msgpack==1.1.0
nest-asyncio==1.6.0
networkx==3.4.2
numpy==2.1.2
opt_einsum==3.4.0
optax==0.2.3
orbax-checkpoint==0.7.0
packaging==24.1
pillow==11.0.0
protobuf==5.28.3
Pygments==2.18.0
pyparsing==3.2.0
python-dateutil==2.9.0.post0
PyYAML==6.0.2
rich==13.9.3
scipy==1.14.1
six==1.16.0
tensorstore==0.1.66
toolz==1.0.0
typing_extensions==4.12.2
zipp==3.20.2
```
You can get all the packages you've installed and their versions with `pip freeze` (the above list is the output from what I installed in my venv in order to run `test_networks.py`).

If you add a `setup.py` file then the install script should look something like
```bash
#!/bin/bash

# Exit script on error
set -e

# Set the virtual environment directory name
VENV_DIR=".venv"

# Step 1: Create the virtual environment
python3 -m venv "$VENV_DIR"

# Step 2: Activate the virtual environment
source "$VENV_DIR/bin/activate"

# Step 3: Upgrade pip and install dependencies from requirements.txt
pip install --upgrade pip

if [ -f "requirements.txt" ]; then
    pip install -r requirements.txt
else
fi

# Step 4: Install the project in editable mode
pip install -e .
```

File: Environment/CTP_generator.py [hunk 3]
```diff
@@ -1,103 +1,153 @@
...
-#With the same number of nodes, the total number of edges can be different.
-#Cannot currently jit this function because of the conversion to numpy array for Delauney
-def generate_graph(n_nodes: int, key: int, use_edge_weights=True, grid_size=10,prop_stoch=0.4
-                   ) -> tuple[jraph.GraphsTuple, int, int]:
+
+# Generate a random graph
+# With the same number of nodes, the total number of edges can be different.
+# Cannot currently jit this function because of the conversion to numpy array for Delauney
+def generate_graph(
+    n_nodes: int, key: int, use_edge_weights=True, grid_size=10, prop_stoch=0.4
```
Comment (alex-schutz) [inline Environment/CTP_generator.py:16]: I know this interface is similar to what I was using, but I think we can make some improvements

- `grid_size` is pretty arbitrary, I forgot to change this for some experiments with n=100 and it took a while to work out why the maps were all the same. We should probably set it to be equal to `n_nodes` or some other more flexible implementation
- Instead of `prop_stoch` the other way to determine the amount of stochastic edges is to specify `k` (number of edges). Ideally we could support both, perhaps they should both default to `None` and throw an error if both or neither are specified

File: Environment/CTP_generator.py [hunk 4]
```diff
@@ -1,103 +1,153 @@
...
+    grid_nodes = jax.vmap(_convert_to_grid, in_axes=(0, None))(node_pos, ymax)
     grid_nodes_jax = jnp.array(grid_nodes).T
 
-    #Choose orign and goal nodes
-    origin,goal = find_goal_and_origin(grid_nodes_jax)
+    # Choose orign and goal nodes
+    origin, goal = find_goal_and_origin(grid_nodes_jax)
 
     # Apply Delauney triangulation to get edges
     delaunay = Delaunay(grid_nodes_jax)
```
Comment (alex-schutz) [inline Environment/CTP_generator.py:30]: It might be possible to write a jittable implementation of Delaunay triangulation (ChatGPT probably), but environment generation isn't a bottleneck at the moment so we'll only cross that bridge if we ever come to it

File: Environment/CTP_generator.py [hunk 5]
```diff
@@ -1,103 +1,153 @@
...
     simplices = delaunay.simplices
     simplices_jax = jnp.array(simplices)
-    
+
     # Extract edges from the simplices, treat them as undirected edges
-    # Remove duplicate edges. Sender node number always > receiver node number now. 
+    # Remove duplicate edges. Sender node number always > receiver node number now.
```
Comment (alex-schutz) [inline Environment/CTP_generator.py:35]: It's probably a good idea to write a class for graphs so that the user doesn't have to remember to always look for the edges with smaller number first. Very simple class like
```python3
class CTPGraph:
    def __init__(self, n: int, edge_weights: dict, edge_probabilities: dict, origin: int, goal: int):
        """
        Initialize the graph with n nodes, edge weights, and edge probabilities.
        
        :param n: Number of nodes in the graph.
        :param edge_weights: Dictionary containing edge weights as {(node1, node2): weight}.
        :param edge_probabilities: Dictionary containing probabilities as {(node1, node2): probability}.
        """
        self.n = n
        self.edge_weights = {self._sorted_edge(edge): weight for edge, weight in edge_weights.items()}
        self.edge_probabilities = {self._sorted_edge(edge): probability for edge, probability in edge_probabilities.items()}
        self.origin = origin
        self.goal = goal
    
    @staticmethod
    def _sorted_edge(edge: tuple) -> tuple:
        """
        Sort the edge tuple to ensure smaller node number comes first.
        
        :param edge: A tuple representing an edge (node1, node2).
        :return: A sorted tuple.
        """
        return tuple(sorted(edge))
    
    @classmethod
    def get_edge_weight(cls, node1: int, node2: int) -> float:
        """
        Get the weight of the edge between node1 and node2, regardless of input order.
        
        :param node1: The first node.
        :param node2: The second node.
        :return: The weight of the edge.
        """
        return cls.edge_weights.get(cls._sorted_edge((node1, node2)), None)
    
    @classmethod
    def get_edge_probability(cls, node1: int, node2: int) -> float:
        """
        Get the probability of the edge between node1 and node2, regardless of input order.
        
        :param node1: The first node.
        :param node2: The second node.
        :return: The probability of the edge.
        """
        return cls.edge_probabilities.get(cls._sorted_edge((node1, node2)), None)
        
    def get_goal()
    def get_origin()
```

I think you're right that we don't actually need to use jraph (unless it's more compatible with jax than using a class like this but I don't immediately see any issues)

File: Environment/CTP_generator.py [hunk 6]
```diff
@@ -1,103 +1,153 @@
...
     all_edges = jnp.concatenate(jax.vmap(_extract_edges)(simplices_jax))
     unique_edges = jnp.unique(all_edges, axis=0)
     senders = unique_edges[:, 0]
     receivers = unique_edges[:, 1]
 
-    #The weight is the Euclidean distance between the nodes, unless use_edge_weights is False
-    edge_weights = jax.lax.cond(use_edge_weights, lambda _:jnp.linalg.norm(grid_nodes_jax[senders] - grid_nodes_jax[receivers], axis=1), lambda _:jnp.ones_like(senders,dtype=jnp.float32),operand=None)
-    
-    blocking_probability = get_blocking_prob(len(senders),key,prop_stoch=prop_stoch)
+    # The weight is the Euclidean distance between the nodes, unless use_edge_weights is False
```
Comment (alex-schutz) [inline Environment/CTP_generator.py:41]: Extension (much) further down the line is to allow the user to input their own metric

File: Environment/CTP_generator.py [hunk 7]
```diff
@@ -1,103 +1,153 @@
...
     is_stochastic_edges = jnp.isin(edge_indices, stoch_edge_idx)
-    edge_probs = jax.vmap(_assign_prob_edge, in_axes=(0,0))(keys,is_stochastic_edges)
+    edge_probs = jax.vmap(_assign_prob_edge, in_axes=(0, 0))(keys, is_stochastic_edges)
     return edge_probs
 
-#Add an edge feature to the graph that stores whether an edge is blocked or not
-def sample_blocking_prob(key:jax.random.PRNGKey,graph:jraph.GraphsTuple) -> jraph.GraphsTuple:
-    #Global context = 1 means the Graph contains the blocking status of all edges
+
+# Add an edge feature to the graph that stores whether an edge is blocked or not
```
Comment (alex-schutz) [inline Environment/CTP_generator.py:126]: I think we could put this within a class to make it clearer that the blockages are drawn from a distribution of all possible graphs. Something like

`CTPGraph`
 - n nodes
 - edge weights
 - edge probabilities
 - `generate_distribution() -> dict[CTPGraphRealisation, float]` (build the entire distribution of possible states, will get very big for large problems)
 - `sample_from _distribution()` (once you have a distribution, sample one realisation)
 - `sample_from_edges()` (directly sample a realisation from the edge probabilities, faster for big problems)

`CTPGraphRealisation`
- contains a reference to parent `CTPGraph` (gives you access to the map itself)
- stores the blocked status (I would add a small data class for the blockage status called e.g. `blockage.BLOCKED` and `blockage.UNBLOCKED` because I keep forgetting whether 1 corresponds to blocked or traversable)
- overrides functions like `get_edge_weight` to only return non-blocked edges
- add functions like `is_solvable`

File: Environment/CTP_environment.py [hunk 8]
```diff
@@ -4,86 +4,111 @@
...
-    #list_of_goals: jnp.array
-    #graph: jraph.GraphsTuple #globals = 0 (does not contain blocking status)
+    # list_of_goals: jnp.array
+    # graph: jraph.GraphsTuple #globals = 0 (does not contain blocking status)
+
 
-#Each observation is a dictionary of dictionaries, where the first key is the agent index and 
-#the second key is the edge (sender,receiver) and the value is the blocking status (0 means not blocked).
+# Each observation is a dictionary of dictionaries, where the first key is the agent index and
+# the second key is the edge (sender,receiver) and the value is the blocking status (0 means not blocked).
```
Comment (alex-schutz) [inline Environment/CTP_environment.py:21]: As we talked about on the call, another (more compact) way of implementing single agent observations is to use an `n`-bit array to store the observation at each node, where the ith bit is the blockage status of the edge between the current node and node `i`. You'll also want to incorporate the observation of the agent's own location.

However, once we extend to the multi-agent case, the observation space will be a lot bigger because communication allows us to observe edges that other agents are observing (and possibly also historical edges), so the labelled-by-edge observations might suit better. Something to think about.

File: Environment/CTP_generator.py (hunk 5 above)
Comment (alex-schutz) [reply Environment/CTP_generator.py:35]: We might want to separate goal/origin selection and graph generation entirely, since in the multi-agent case the agents will have different goals but the same underlying graph

File: Environment/CTP_environment.py [hunk 9]
```diff
@@ -4,86 +4,111 @@
...
-            self.true_graph = CTP_generator.add_expensive_edge(self.true_graph,largest_edge_weight*3,self.origin,self.goal)
+        self.true_graph = CTP_generator.sample_blocking_prob(key, self.agent_graph)
+        self.solvable = CTP_generator.is_solvable(
+            self.true_graph, self.origin.item(), self.goal.item()
+        )
+
+        if self.solvable is False and self.add_expensive_edge is True:
+            # Add an expensive edge.
+            largest_edge_weight = jnp.max(self.true_graph.edges["weight"])
+            self.true_graph = CTP_generator.add_expensive_edge(
```
Comment (alex-schutz) [inline Environment/CTP_environment.py:72]: I think the expensive edge should always be there even if the graph is solvable without, because it should be a valid policy for the agent to just take that edge in whichever realisation it exists in

File: Environment/CTP_environment.py [hunk 10]
```diff
@@ -4,86 +4,111 @@
...
+        self, state: EnvState, action: jnp.array
+    ) -> tuple[Observation, EnvState, int, bool]:
         # Return observation, state, reward, and whether the episode is done
         # This if statement is not JAX-compatible (preventing tracing because of boolean conversion error)
-        if(self.add_expensive_edge is False and action == self.true_graph.n_node):
-            result = jax.lax.cond(self.solvable,lambda _: (0, 0, -5000, True),lambda _: (0, 0, 0, True),operand=None)
+        if self.add_expensive_edge is False and action == self.true_graph.n_node:
+            result = jax.lax.cond(
+                self.solvable,
+                lambda _: (0, 0, -5000, True),
```
Comment (alex-schutz) [inline Environment/CTP_environment.py:89]: We should make this number configurable, as with any of the reward values

File: Environment/CTP_environment.py [hunk 11]
```diff
@@ -94,28 +119,67 @@ def step(self, state:EnvState,action:jnp.array) -> tuple[Observation,EnvState,in
...
+                                    edge_indices[0][i]
+                                ]
+                            )  # Converting the edge status to int as well
+                            for i in range(len(edge_indices[0]))
+                        }
+                    )
+                }
+            )
         return observation
-
```
Comment (alex-schutz) [inline Environment/CTP_environment.py:118]: The environment is definitely something that should be unit tested. Make some small maps and check that the actions do what you expect them to do, and return the correct reward and observations

**Review by alex-schutz — COMMENTED** [review]
Some comments on the environment design.
I think a lot of this code will be made easier to maintain by using some classes/containers

**Comment by alex-schutz** [comment]
It might feel like we're spending a lot of time on the boring early stuff, but if we get this part right then the rest of the code will be much easier to maintain and also understand. It's worth getting it right now while the project is small
